File: dart/lib/bebop_handlers/init_state.dart
import 'package:lean_dart_client/classes/game_state.dart';
import 'package:lean_dart_client/classes/player.dart';
import 'bebop_handler.dart';
import 'package:lean_dart_client/schemas.g.dart';

class InitStateMsgHandler extends BebopHandler {
  @override
  void handleMessage(sender, message) {
    var player = sender as Player;
    var decodedMessage = InitState.decode(message);

    for (var pos in decodedMessage.playerInitPositions.entries) {
      GameState.instance.playerState.playerPositions
          .update(pos.key, (value) => pos.value, ifAbsent: () => pos.value);
    }

    for (var lives in decodedMessage.playerInitLives.entries) {
      GameState.instance.playerState.lives.update(
          lives.key, (value) => lives.value,
          ifAbsent: () => lives.value);
    }

    for (var score in decodedMessage.playerInitLives.entries) {
      GameState.instance.playerState.scores.update(
          score.key, (value) => score.value,
          ifAbsent: () => score.value);
    }

    for (var ghost in decodedMessage.ghostInitPositions.entries) {
      GameState.instance.ghostPositions.update(
          ghost.key, (value) => ghost.value,
          ifAbsent: () => ghost.value);
    }

    player.position = player.startPosition =
        decodedMessage.playerInitPositions[GameState.instance.session.clientId];

    GameState.instance.scorePointPositions =
        decodedMessage.scorePointInitPositions;

    GameState.instance.completeInitFuture();
  }
}

File: typescript/src/bebop_handlers/initState.ts
import GameState from '../classes/gameState';
import Player from '../classes/player';
import { InitState } from '../models/pactheman.models';
import PositionExtended from '../utils/extensions/positionExtensions';
import BebopHandler from './bebopHandler';

export default class InitStateMsgHandler implements BebopHandler {
    handleMessage(sender: any, message: any): void {
        let player = sender as Player;
        message = InitState.decode(message);

        for (var pos of message.playerInitPositions.entries()) {
            GameState.Instance.playerState.playerPositions.set(pos[0], pos[1]);
          }
      
          for (var lives of message.playerInitLives.entries()) {
            GameState.Instance.playerState.lives.set(lives[0], lives[1]);
          }
      
          for (var score of message.playerInitLives.entries()) {
            GameState.Instance.playerState.scores.set(score[0], score[1]);
          }
      
          for (var ghost of message.ghostInitPositions.entries()) {
            GameState.Instance.ghostPositions.set(ghost[0], ghost[1]);
          }
      
          player.position = player.startPosition =
              PositionExtended.fromPosition(message.playerInitPositions.get(GameState.Instance.session.clientId ?? "") ?? { x: 0, y: 0 });
      
          GameState.Instance.scorePointPositions =
              message.scorePointInitPositions;
      
          GameState.Instance.completeInitFuture();
    }
}